src/vm/runners/cairo_runner.zig

```diff
@@ -216,39 +222,81 @@ pub const CairoRunner = struct {
         // but to situate the functionality with Cairo's read-only memory,
         // the frame pointer register is used to point to the current frame in the stack
         // the runner sets the return fp and establishes the end address that execution treats as the endpoint.
-        try self.function_call_stack.append(MaybeRelocatable.fromRelocatable(return_fp));
-        try self.function_call_stack.append(MaybeRelocatable.fromRelocatable(end));
+        try stack.append(MaybeRelocatable.fromRelocatable(return_fp));
+        try stack.append(MaybeRelocatable.fromRelocatable(end));
 
         self.initial_fp = self.execution_base;
-        self.initial_fp.addUintInPlace(@as(u64, self.function_call_stack.items.len));
+        self.initial_fp.addUintInPlace(@as(u64, stack.items.len));
         self.initial_ap = self.initial_fp;
 
         self.final_pc = &end;
-        try self.initState(entrypoint);
+        try self.initState(entrypoint, stack);
         return end;
     }
 
     /// Initializes runner state for execution of a program from the `main()` entrypoint.
     pub fn initMainEntrypoint(self: *Self) !Relocatable {
+        var stack = std.ArrayList(MaybeRelocatable).init(self.allocator);
+        defer stack.deinit();
+
         for (self.vm.builtin_runners.items) |*builtin_runner| {
             const builtin_stack = try builtin_runner.initialStack(self.allocator);
             defer builtin_stack.deinit();
             for (builtin_stack.items) |item| {
-                try self.function_call_stack.append(item);
+                try stack.append(item);
             }
         }
-        // TODO handle the case where we are running in proof mode
+
+        if (self.isProofMode()) {
+            var target_offset: usize = 2;
+
+            if (self.runner_mode == .proof_mode_canonical) {
+                var stack_prefix = try std.ArrayList(MaybeRelocatable).initCapacity(self.allocator, 2 + stack.items.len);
+                defer stack_prefix.deinit();
+
+                try stack_prefix.append(MaybeRelocatable.fromRelocatable(try self.execution_base.addUint(target_offset)));
+                try stack_prefix.appendSlice(stack.items);
+
+                var execution_public_memory = try std.ArrayList(usize).initCapacity(self.allocator, stack_prefix.items.len);
+                for (0..stack_prefix.items.len) |v| {
+                    try execution_public_memory.append(v);
+                }
+                self.execution_public_memory = execution_public_memory;
+
+                try self.initState(try (if (self.program.getStartPc()) |start| start else RunnerError.NoProgramStart), &stack_prefix);
```

Comment:
Should be 

```zig
try self.initState(self.program.getStartPc() orelse return RunnerError.NoProgramStart, &stack_prefix);
```

no?

Reply:
yes good solution i forgot about orelse, but circle brackets we need anyway :(